driver-core/src/main/com/mongodb/internal/operation/AsyncCommandBatchCursor.java

```diff
@@ -228,187 +170,149 @@ public int getMaxWireVersion() {
+                        callback.onResult(nextBatch, null);
+                        return;
+                    }
+
+                    if (!resourceManager.operable()) {
+                        callback.onResult(emptyList(), null);
+                        return;
+                    }
+
+                    getMoreLoop(connection, commandCursorResult.getServerCursor(), callback);
+        });
     }
 
-    private void getMore(final AsyncConnection connection, final ServerCursor cursor, final SingleResultCallback<List<T>> callback) {
-        connection.commandAsync(namespace.getDatabaseName(), asGetMoreCommandDocument(cursor.getId(), connection.getDescription()),
-                NO_OP_FIELD_NAME_VALIDATOR, ReadPreference.primary(), CommandResultDocumentCodec.create(decoder, "nextBatch"),
-                connectionSource.getOperationContext(), new CommandResultSingleResultCallback(connection, cursor, callback));
-   }
+    private CommandCursorResult<T> toCommandCursorResult(final ServerAddress serverAddress, final String fieldNameContainingBatch,
+            final BsonDocument commandCursorDocument) {
+        CommandCursorResult<T> commandCursorResult = new CommandCursorResult<>(serverAddress, fieldNameContainingBatch,
+                commandCursorDocument);
+        logCommandCursorResult(commandCursorResult);
+        return commandCursorResult;
+    }
 
-    private BsonDocument asGetMoreCommandDocument(final long cursorId, final ConnectionDescription connectionDescription) {
-        BsonDocument document = new BsonDocument("getMore", new BsonInt64(cursorId))
-                .append("collection", new BsonString(namespace.getCollectionName()));
+    @ThreadSafe
+    private static final class ResourceManager extends CursorResourceManager<AsyncConnectionSource, AsyncConnection> {
 
-        int batchSizeForGetMoreCommand = Math.abs(getNumberToReturn(limit, this.batchSize, count.get()));
-        if (batchSizeForGetMoreCommand != 0) {
-            document.append("batchSize", new BsonInt32(batchSizeForGetMoreCommand));
+        ResourceManager(
+                final MongoNamespace namespace,
+                final AsyncConnectionSource connectionSource,
+                @Nullable final AsyncConnection connectionToPin,
+                @Nullable final ServerCursor serverCursor) {
+            super(namespace, connectionSource, connectionToPin, serverCursor);
         }
-        if (maxTimeMS != 0) {
-            document.append("maxTimeMS", new BsonInt64(maxTimeMS));
+
+        /**
+         * Thread-safe.
+         * Executes {@code operation} within the {@link #tryStartOperation()}/{@link #endOperation()} bounds.
+         */
+        <R> AsyncCallbackSupplier<R> execute(final AsyncCallbackSupplier<R> callbackSupplier) {
+            return callback -> {
+                if (!tryStartOperation()) {
+                    callback.onResult(null, new IllegalStateException(MESSAGE_IF_CONCURRENT_OPERATION));
```

Comment:
`tryStartOperation` returns `false` iff the cursor is closed, which means that `MESSAGE_IF_CONCURRENT_OPERATION` is improper here. `CommandBatchCursor` uses `MESSAGE_IF_CLOSED_AS_CURSOR` in this situation.